### PR TITLE
detect improper use of t.throws

### PR DESCRIPTION
**Pull request by jamestalmage**
Protects against a common misuse of t.throws (Like that seen in #739).

This required the creation of a custom babel plugin.

https://github.com/jamestalmage/babel-plugin-ava-throws-helper

Sample output:

```
Improper usage of t.throws detected at /Users/jamestalmage/dev/ava/test/fixture/improper-t-throws.js (4:10).

You should wrap the following expression in a function:

  throwSync()

Like this:

  function() {
    throwSync()
  }

See https://github.com/sindresorhus/ava#throwsfunctionpromise-error-message for more details.
```

The linked documentation isn't that helpful. We should probably make a short recipe about this. (maybe a `common-mistakes` recipe? This seems like too small a topic for a whole recipe). We can update the URL when that recipe is written.

### CODE REVIEW

File: lib/test.js [hunk 1]
```diff
@@ -68,6 +68,24 @@ Test.prototype._assert = function (promise) {
...
 Test.prototype._setAssertError = function (err) {
+	var data = err._avaThrowsHelperData;
+	if (data) {
+		console.error(
+			[
+				'Improper usage of t.throws detected at %s (%d:%d).',
+				'You should wrap the following expression in a function:',
+				'  %s',
+				'Like this:',
+				'  function() {\n    %s\n  }',
```
Comment (novemberborn) [inline lib/test.js]: Promote arrow functions?

File: lib/test.js [hunk 2]
```diff
@@ -68,6 +68,24 @@ Test.prototype._assert = function (promise) {
 };
 
 Test.prototype._setAssertError = function (err) {
+	var data = err._avaThrowsHelperData;
```
Comment (novemberborn) [inline lib/test.js]: Probably need to guard against `err` being `null` or `undefined` here?

File: test/caching-precompiler.js [hunk 3]
```diff
@@ -145,7 +146,7 @@ test('uses babelConfig for babel options when babelConfig is an object', functio
 	t.true('inputSourceMap' in options);
 	t.false(options.babelrc);
 	t.same(options.presets, ['stage-2', 'es2015']);
-	t.same(options.plugins, [customPlugin, powerAssert, rewrite, transformRuntime]);
+	t.same(options.plugins, [customPlugin, powerAssert, throwsHelper, rewrite, transformRuntime]);
```
Comment (novemberborn) [inline test/caching-precompiler.js]: Without checking the `power-assert` transformations, are there cases the Babel plugin can no longer catch?

**Comment by novemberborn** [comment]
> The linked documentation isn't that helpful. We should probably make a short recipe about this

:+1: 



> (maybe a common-mistakes recipe?

#404?

File: test/caching-precompiler.js (hunk 3 above)
Comment (jamestalmage) [inline test/caching-precompiler.js]: Shouldn't be. The `babel-throws-helper` wrapper is totally transparent to the code it wraps.

File: lib/test.js (hunk 1 above)
Comment (jamestalmage) [inline lib/test.js]: I considered that.

But if the user is green enough that that they don't fully understand how `throw` works, I was concerned about piling on another abstraction they need to understand.

File: test/caching-precompiler.js (hunk 3 above)
Comment (novemberborn) [inline test/caching-precompiler.js]: Sure, but `power-assert` might rewrite the `t.throws()` argument expression such that it is computed outside of `t.throws()`. Again, I don't know if it does, but if it did then you wouldn't be able to catch it.

File: lib/test.js (hunk 1 above)
Comment (novemberborn) [inline lib/test.js]: Yea fair enough.

**Comment by vdemedes** [comment]
> /Users/jamestalmage/dev/ava/test/fixture/improper-t-throws.js (4:10)

Perhaps displaying relative path would be better and would align with our "clean stack traces" feature.

**Comment by jamestalmage** [comment]
Updated with relative file paths and colors:

screenshot 2016-04-10 16 44 14

File: test/caching-precompiler.js (hunk 3 above)
Comment (jamestalmage) [inline test/caching-precompiler.js]: Oh - Understood. Still not a problem though. `power-assert` wraps inline, just like `ava-throws-helper`.

```js
t.true(_rec._expr(_rec._capt(_rec._capt(a, 'arguments/0/left') === 'bar', 'arguments/0'), {
	content: 't.true(a === \'bar\')',
	filepath: 'test/fixture/power-assert.js',
	line: 6
}));
```

Also, we don't enhance `t.throws` with `power-assert`.

**Comment by vdemedes** [comment]
I know it is an overkill, but would be incredibly cool to have syntax highlight in errors (not just this one).

File: test/caching-precompiler.js (hunk 3 above)
Comment (twada) [inline test/caching-precompiler.js]: > we don't enhance t.throws with power-assert.

Right. `t.throws` is [listed as NON_ENHANCED_PATTERNS here](https://github.com/sindresorhus/ava/blob/84c05fe98ee5887e49262d48f1134cdf19118a7f/lib/enhance-assert.js#L23-L29). power-assert doesn't touch its arguments.

**Comment by jamestalmage** [comment]
> I know it is an overkill, but would be incredibly cool to have syntax highlight in errors (not just this one).

Like this?:

screenshot 2016-04-11 01 49 24

**Comment by jamestalmage** [comment]
I think maybe the middle section:

```
The following expression:
  t.throws()
```

Is kinda useless with the codeframe right above.

**Comment by jamestalmage** [comment]
screenshot 2016-04-11 02 03 30

**Comment by vdemedes** [comment]
Super awesome!

**Comment by sindresorhus** [comment]
Marvelous. :shipit:

**Comment by novemberborn** [comment]
Cool!

Should update the default plugin sections in [the readme](https://github.com/sindresorhus/ava/tree/195390ec1db90cf7d394407c846a69fbaa08806d#es2015-support) and [babelrc recipe](https://github.com/sindresorhus/ava/blob/195390ec1db90cf7d394407c846a69fbaa08806d/docs/recipes/babelrc.md#notes).

**Comment by jamestalmage** [comment]
> Should update the default plugin sections in the readme and babelrc recipe.

I went with just the `babelrc` recipe. I feel like most user's don't care, so I don't want to pollute the readme with every side-effect free plugin we install. Really, we could just update the readme to say. "AVA always includes a few internal plugins regardless of configuration, but they should not impact the behavior of hour code", and then link to the Notes section of the `babelrc` recipe.

**Comment by sindresorhus** [comment]
> "AVA always includes a few internal plugins regardless of configuration, but they should not impact the behavior of hour code", and then link to the Notes section of the babelrc recipe.

:+1:

**Comment by jfmengels** [comment]
:tada:

**Comment by sindresorhus** [comment]
🍾✨

**Pull request merged**